nano/node/confirming_set.cpp

```diff
@@ -0,0 +1,120 @@
+#include <nano/lib/thread_roles.hpp>
+#include <nano/node/confirming_set.hpp>
+#include <nano/node/write_database_queue.hpp>
+#include <nano/secure/ledger.hpp>
+#include <nano/store/component.hpp>
+
+nano::confirming_set::confirming_set (nano::ledger & ledger, nano::write_database_queue & write_queue, std::chrono::milliseconds batch_time) :
+	ledger{ ledger },
+	write_queue{ write_queue },
+	batch_time{ batch_time }
+{
+}
+
+nano::confirming_set::~confirming_set ()
+{
+	debug_assert (!thread.joinable ());
+}
+
+void nano::confirming_set::add (nano::block_hash const & hash)
+{
+	std::lock_guard lock{ mutex };
+	set.insert (hash);
+	condition.notify_all ();
+}
+
+void nano::confirming_set::start ()
+{
+	thread = std::thread{ [this] () { run (); } };
+}
+
+void nano::confirming_set::stop ()
+{
+	{
+		std::lock_guard lock{ mutex };
+		stopped = true;
+		condition.notify_all ();
+	}
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+bool nano::confirming_set::exists (nano::block_hash const & hash) const
+{
+	std::lock_guard lock{ mutex };
+	return set.count (hash) != 0 || processing.count (hash) != 0;
+}
+
+std::size_t nano::confirming_set::size () const
+{
+	std::lock_guard lock{ mutex };
+	return set.size () + processing.size ();
+}
+
+void nano::confirming_set::run ()
+{
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	std::unique_lock lock{ mutex };
+	// Run the confirmation loop until stopped
+	while (!stopped)
+	{
+		condition.wait (lock, [&] () { return !set.empty () || stopped; });
+		// Loop if there are items to process
+		if (!stopped && !set.empty ())
+		{
+			std::deque<std::shared_ptr<nano::block>> cemented;
+			std::deque<nano::block_hash> already;
+			// Move items in to back buffer and release lock so more items can be added to the front buffer
+			processing = std::move (this->set);
+			// Process all items in the back buffer
+			for (auto i = processing.begin (), n = processing.end (); !stopped && i != n;)
+			{
+				lock.unlock (); // Waiting for db write is potentially slow
+				write_queue.wait (nano::writer::confirmation_height);
```

Comment:
Good catch, that wasn't clear from the description. I opened https://github.com/nanocurrency/nano-node/pull/4522 to add [[nodiscard]] to help avoid this in the future.

Reply:
It might be worth to consider whether this write queue is needed at all, from what I understand a single database write transaction can be active at a time anyway.